File: ProjectCS/UI/DisciplineListPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SessionResultsCS.UI
{
    public class DisciplineListPage : LoopPage
    {
        private Speciality _speciality;
        private int _course;

        private DisciplineList _disciplineList;
        private MenuActionPicker _picker;


        public DisciplineListPage(Speciality speciality, int course)
        {
            _speciality = speciality;
            _course = course;
            _disciplineList = speciality.GetDisciplineList(_course);
            _picker = new MenuActionPicker([
                ("Изменить название дисциплины", ChangeDisciplineName),
                ("Добавить дисциплину", AddDiscipline),
                ("Удалить дисциплину", DeleteDiscipline),
                ("Назад", StopLoop),
            ]);
        }

        protected override void Loop()
        {
            Console.WriteLine($"==== Дисциплины - {_course} курс ====");
            PrintDisciplines();
            Console.WriteLine("Действия:");
            _picker.PickAndRun();
        }


        private void ChangeDisciplineName()
        {
            int selectedIndex = SelectDiscipline();
            if (selectedIndex != -1)
            {
                Discipline oldDiscipline = _disciplineList.GetDisciplineAt(selectedIndex);

                try
                {
                    Console.Write("Введите новое название дисциплины: ");
                    string name = Console.ReadLine()!;

                    Discipline newDiscipline = new Discipline(name, oldDiscipline.AttestationType);

                    _disciplineList.RemoveDiscipline(oldDiscipline);
                    _disciplineList.AddDiscipline(newDiscipline);

                    foreach (Group group in _speciality.GetGroups().Where(group => group.Course == _course))
                    {
                        foreach (Student student in group.GetStudents())
                        {
                            AttestationResult? result = student.SessionResults.GetResult(oldDiscipline);
                            if (result != null)
                            {
                                student.SessionResults.DeleteResult(oldDiscipline);
                                student.SessionResults.SetResult(newDiscipline, result);
                            }
                        } 
                    }
                }
                catch (ArgumentException)
                {
                    Console.WriteLine("Не удалось изменить дисциплину");
                }
            }
        }

        private void AddDiscipline()
        {
            Console.Write("Введите название новой дисциплины: ");
            string name = Console.ReadLine()!;
            Console.WriteLine("Выберите тип аттестации новой дисциплины: ");
            AttestationType attestationType = PickAttestationType();

            try
            {
                Discipline discipline = new Discipline(name, attestationType);
                _disciplineList.AddDiscipline(discipline);
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Не удалось добавить дисциплину");
            }
        }

        private void DeleteDiscipline()
        {
            int selectedIndex = SelectDiscipline();
            if (selectedIndex != -1)
            {
                Discipline deletedDiscipline = _disciplineList.GetDisciplineAt(selectedIndex);

                foreach (Group group in _speciality.GetGroups().Where(group => group.Course == _course))
                {
                    foreach (Student student in group.GetStudents())
                    {
                        student.SessionResults.DeleteResult(deletedDiscipline);
                    }
                }

                _disciplineList.RemoveDisciplineAt(selectedIndex);
            }
        }


        private void PrintDisciplines()
        {
            foreach (var (index, discipline) in _disciplineList.Index())
            {
                Console.WriteLine($"{index + 1}. {discipline}");
            }
        }

        private AttestationType PickAttestationType()
        {
            QuickItemPicker attestationTypePicker = new QuickItemPicker(["Экзамен", "Зачёт"]);
            int selected = attestationTypePicker.Pick();
            switch (selected)
            {
                case 0: return AttestationType.Exam;
                case 1: return AttestationType.PassFailExam;
                default: return AttestationType.Exam;
            }
        }

        private int SelectDiscipline()
        {
            Console.Write("Введите номер дисциплины: ");
            string input = Console.ReadLine()!;
            if (int.TryParse(input, out int value) && (value >= 1 && value <= _disciplineList.GetSize()))
            {
                return value - 1;
            }
            else
            {
                Console.WriteLine("Введён неверный номер");
                return -1;
            }
        }
    }
}

File: Project/UI/DisciplineListPage.cpp
#include "UI/DisciplineListPage.hpp"

#include <format>
#include <ranges>
#include <algorithm>


DisciplineListPage::DisciplineListPage(std::shared_ptr<Speciality> speciality, int course) :
	m_speciality(speciality),
	m_disciplineList(speciality->GetDisciplineList(course)),
	m_course(course) 
{
	m_picker = std::make_unique<MenuActionPicker>(std::vector<MenuActionPicker::Item>{
		{ "Изменить название дисциплины", [this]() { ChangeDisciplineName(); } },
		{ "Добавить дисциплину", [this]() { AddDiscipline(); } },
		{ "Удалить дисциплину", [this]() { DeleteDiscipline(); } },
		{ "Назад", [this]() { StopLoop(); } }
	});
}

void DisciplineListPage::Loop() {
	std::cout << std::format("==== Дисциплины - {} курс ====\n", m_course);
	PrintDisciplines();
	std::cout << "Действия:\n";
	m_picker->PickAndRun();
}

void DisciplineListPage::ChangeDisciplineName() {
	int selectedIndex = SelectDiscipline();
	if (selectedIndex != -1) {
		std::shared_ptr<Discipline> oldDiscipline = m_disciplineList.GetDisciplineAt(selectedIndex);

		try {
			std::cout << "Введите новое название дисциплины: ";
			std::string name;
			std::getline(std::cin, name);

			std::shared_ptr<Discipline> newDiscipline = std::make_shared<Discipline>(name, oldDiscipline->GetAttestationType());

			m_disciplineList.RemoveDiscipline(oldDiscipline);
			m_disciplineList.AddDiscipline(newDiscipline);

			for (auto& group : m_speciality->GetGroups() | std::views::filter([&](auto& group) { return group->GetCourse() == m_course; })) {
				for (auto& student : group->GetStudents()) {
					std::shared_ptr<AttestationResult> result = student->GetSessionResults().GetResult(oldDiscipline);
					if (result != nullptr) {
						student->GetSessionResults().DeleteResult(oldDiscipline);
						student->GetSessionResults().SetResult(newDiscipline, result);
					}
				}
			}
		} catch (std::invalid_argument) {
			std::cout << "Не удалось изменить дисциплину\n";
		}
	}
}

void DisciplineListPage::AddDiscipline() {
	std::cout << "Введите название новой дисциплины: ";
	std::string name;
	std::getline(std::cin, name);

	std::cout << "Выберите тип аттестации новой дисциплины:\n";
	Discipline::AttestationType attestationType = PickAttestationType();

	try {
		m_disciplineList.AddDiscipline(std::make_shared<Discipline>(name, attestationType));
	} catch (std::invalid_argument) {
		std::cout << "Не удалось добавить дисциплину\n";
	}
}

void DisciplineListPage::DeleteDiscipline() {
	int selectedIndex = SelectDiscipline();
	if (selectedIndex != -1) {
		std::shared_ptr<Discipline> deletedDiscipline = m_disciplineList.GetDisciplineAt(selectedIndex);

		for (auto& group : m_speciality->GetGroups() | std::views::filter([&](auto& group) { return group->GetCourse() == m_course; })) {
			for (auto& student : group->GetStudents()) {
				student->GetSessionResults().DeleteResult(deletedDiscipline);
			}
		}

		m_disciplineList.RemoveDisciplineAt(selectedIndex);
	}
}

void DisciplineListPage::PrintDisciplines() {
	for (int i = 0; i < m_disciplineList.GetSize(); i++) {
		std::cout << std::format("{}. {}\n", i + 1, m_disciplineList.GetDisciplineAt(i)->ToString());
	}
}

Discipline::AttestationType DisciplineListPage::PickAttestationType() {
	QuickItemPicker attestationTypePicker({"Экзамен", "Зачёт"});
	int selected = attestationTypePicker.Pick();
	switch (selected) {
		case 0: return Discipline::AttestationType::eExam;
		case 1: return Discipline::AttestationType::ePassFailExam;
	}
}

int DisciplineListPage::SelectDiscipline() {
	int value;
	std::cout << "Введите номер дисциплины: ";
	if ((std::cin >> value) && (value >= 1 && value <= m_disciplineList.GetSize())) {
		std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
		return value - 1;
	} else {
		std::cin.clear();
		std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
		std::cout << "Введён неверный номер\n";
		return -1;
	}
}